matching-notification-api/src/main/java/org/phenotips/matchingnotification/internal/MatchesByPatient.java

```diff
@@ -0,0 +1,503 @@
+     *    internalMap.get("p1").get("p2") in addition to m1 and m2.
+     *
+     * The set where matches are saved is ordered, and matches where the local patient is the matched patient come
+     * before matches where it is the reference. This way, when creating a return value collection for
+     * getMatchesForLocalPatientId(id, true), matches where id is matched are preferred. That is, if m1 and m2 are
+     * equivalent and m1.getMatchedPatientId()==id then m1 is returned and m2 is not.
+     */
+    private Map<String, Map<String, Set<PatientMatch>>> internalMap;
+
+    private int size;
+
+    /**
+     * @version $Id$
+     */
+    public class MatchesByPatientIterator implements Iterator<PatientMatch>
+    {
+        private MatchesByPatient mbp;
+
+        private Iterator<Map<String, Set<PatientMatch>>> primaryKeyIterator;
+        private Iterator<Set<PatientMatch>> secondaryKeyIterator;
+        private Iterator<PatientMatch> setIterator;
+
+        private PatientMatch returnNext;
+        private PatientMatch returnedLast;
+
+        private Set<PatientMatch> returnedSet;
+
+        private boolean skipEquivalents;
+
+        /**
+         * @param skipEquivalents skip any match if its equivalent was retrieved
+         * @param mbp MatchesByPatient object that this iterator iterates over
+         */
+        public MatchesByPatientIterator(MatchesByPatient mbp, boolean skipEquivalents)
+        {
+            this.mbp = mbp;
+            this.skipEquivalents = skipEquivalents;
+            this.returnedSet = new HashSet<>();
+            this.initialize();
+        }
+
+        private void initialize()
+        {
+            this.primaryKeyIterator = this.mbp.internalMap.values().iterator();
+            if (this.primaryKeyIterator.hasNext()) {
+                this.secondaryKeyIterator = this.primaryKeyIterator.next().values().iterator();
+                if (this.secondaryKeyIterator.hasNext()) {
+                    this.setIterator = this.secondaryKeyIterator.next().iterator();
+                    if (this.setIterator.hasNext()) {
+                        this.returnNext = this.setIterator.next();
+                    }
+                }
+            }
+        }
+
+        @Override
+        public boolean hasNext()
+        {
+            return this.returnNext != null;
+        }
+
+        @Override
+        public PatientMatch next()
+        {
+            this.returnedLast = this.returnNext;
+
+            this.returnedSet.add(returnedLast);
+            boolean keepLooking = true;
+            while (keepLooking) {
+                this.moveToNext();
+
+                keepLooking = false;
+                if (this.returnNext != null) {
+                    if (this.returnedSet.contains(this.returnNext)) {
+                        keepLooking = true;
+                    }
+                    if (this.skipEquivalents && containsEquivalent(returnedSet, this.returnNext)) {
```

Comment:
It can, but I think this is more readable - 128 checks if the candidate for returning was already returned, 131 checks if we want to skip equivalent matches and an equivalent match was returned.